controller.go

```diff
@@ -469,3 +475,79 @@ func upgradePVFrom12(pv *PV) (deleted bool, err error) {
 	}
 	return false, nil
 }
+
+func provisionClaim(pvc *api.PersistentVolumeClaim) error {
+	// Provisioning can take some time. Start it in a separate goroutine.
+	// Keep a map of running provisioners and start new provisioning only if
+	// it's not already running. There will be some locks around, these are
+	// left out to implementation.
+	//
+
+	if !runningProvisioners.isRunning(pvc) {
+		runningProvisioners.markAsRunning(pvc)
+		go doProvisionClaim(pvc)
+	}
+	return nil
+}
+
+func doProvisionClaim(pvc *api.PersistentVolumeClaim) {
+	// On exit, mark the provisioner for this claim as finished. We don't want
+	// multiple doProvisionClaims running in parallel for the same claim.
+	defer runningProvisioners.markAsFinished(pvc)
+
+	// We're the only provisioner running for the PVC now. A previous
+	// doProvisionClaim may just have finished while we were waiting for the
+	// locks. Check that PV (with deterministic name) hasn't been provisioned
+	// yet.
+	pvName := calculateProvisionedPVNameForClaim(claim)
+	pv := kubeClient.Core().PersistentVolumes().Get(pvName)
+	if pv != nil {
+		// Previous doProvisionClaim has provisioned something, let the usual
+		// binding happen.
+		return
+	}
+
+	// Now we are sure that we need to provision something.
+	plugin := findProvisionerPluginForPV(claim)
+	if plugin = nil {
+		// Emit event that provisioning failed.
+		// Do not modify the claim and let the next syncClaim() try again.
+		return
+	}
+
+	// Provision the volume. Plugin is responsible to return valid PV object.
+	pv, err := plung.Provision(claim)
+	if err != nil {
+		// Emit event that provisioning failed.
+		// Do not modify the claim and let the next syncClaim() try again.
+		return
+	}
+
+	// NOTE: if the controller crashes here (or exits or gets SIGTERM), we will
+	// leave orphaned storage asset without having a PV for it in Kubernetes.
+
+	// Make sure the PV has the right name.
+	pv.Name = calculateProvisionedPVNameForClaim(claim)
+	// Bind it to the claim
+	pv.Spec.ClaimPtr.Name = claim.Name
+	pv.Spec.ClaimPtr.Namespace = claim.Namespace
+	// Make sure we delete the PV when the claim is deleted or bound to a
+	// different PV. Setting UID is part of it (see syncPV).
+	pv.Spec.ClaimPtr.UID = claim.UID
+	pv.Spec.ReclaimPolicy = api.ReclaimPolicyDelete
+
+	// Save it in etcd. TODO: try several times.
+	if err = kubeClient.Core().PersistentVolumes().Add(pv); err != nil {
+		// Save failed. Now we have a storage asset outside of Kubernetes,
+		// but we don't have appropriate PV object for it.
+		// Emit some event here and do our best to delete the storage asset.
+		deleteVolume(pv)
```

Comment:
I guess you could always have a misbehaving proxy or something between you and the API server, so this is probably safer.